x-pack/plugins/profiling/common/callercallee.ts

```diff
@@ -272,6 +282,7 @@ export function fromCallerCalleeIntermediateNode(
   root: CallerCalleeIntermediateNode
 ): CallerCalleeNode {
   const node = createCallerCalleeNode({
+    ID: root.id,
```

Comment:
Could you help me understand the algorithm used to compare two flamegraphs?

Reviewing our prior implementation, it appears we walk the two flamegraphs at the same time and then we use the respective serialized frame group (defined by `hashFrameGroup` in this PR) to compare each visited node. The prior flamegraphs use a data structure similar to `CallerCalleeNode`.

I think the ID generated here using `object_hash` might be causing unexpected results, but I haven't done a full walkthrough. Based on local experiments, we may be able to replace `ID: root.id` with `ID: hashFrameGroup(root.frameGroup)`. Then we can remove `CallerCalleeIntermediateNode.id`, `StackFrameMetadata.FrameID`, and `object_hash`.

Reply:
@jbcrail I think my assumption here is that hashFrameGroup creates a unique ID per frame, but not per frame per position in its stack trace, if that makes sense? I recall using the frame group hash and ending up with duplicates, too. But maybe I'm misremembering.

Reply:
@dgieselaar Yep, that sounds right. One difference between the prior flamegraph and the new flamegraph (including their respective differential flamegraphs) is that the prior implementation ingests a tree-like call graph, whereas we now use a flattened columnar format. We lose the implicit stack trace position available in the tree-like format so the ID generated by  `hashFrameGroup` isn't enough for us to compare call graphs in the new format, thus, we needed a different ID. It look me some time to refresh my memory about the prior implementation and walk through the differences.

Just from reading the code, I am not sure what impact using `object_hash` and adding another field to `StackFrameMetadata` will have on the overall latency (`StackFrameMetadata` is returned by the traces view and Top N functions). However, I don't think it's a blocker for this PR.

FYI: after this PR is merged and if time permits, I will likely do another API performance optimization pass. Our stretch goal is to achieve sub-1-second latencies.